precompiles/assets/types.go

```diff
@@ -32,27 +35,132 @@ func (p Precompile) GetDepositWithdrawParamsFromInputs(ctx sdk.Context, args []i
 	}
 	depositWithdrawParams.StakerAddress = stakerAddr[:clientChainAddrLength]
 
 	opAmount, ok := args[3].(*big.Int)
 	if !ok || opAmount == nil || opAmount.Cmp(big.NewInt(0)) == 0 {
-		return nil, xerrors.Errorf(exocmn.ErrContractInputParaOrType, 3, "*big.Int", opAmount)
+		return nil, xerrors.Errorf(exocmn.ErrContractInputParaOrType, 3, "*big.Int", args[3])
 	}
 	depositWithdrawParams.OpAmount = sdkmath.NewIntFromBigInt(opAmount)
 
 	return depositWithdrawParams, nil
 }
+
+func (p Precompile) ClientChainInfoFromInputs(_ sdk.Context, args []interface{}) (*types.ClientChainInfo, error) {
+	if len(args) != 5 {
+		return nil, xerrors.Errorf(cmn.ErrInvalidNumberOfArgs, 5, len(args))
+	}
+	clientChain := types.ClientChainInfo{}
+	clientChainID, ok := args[0].(uint32)
+	if !ok {
+		return nil, fmt.Errorf(exocmn.ErrContractInputParaOrType, 0, "uint32", args[0])
+	}
+	clientChain.LayerZeroChainID = uint64(clientChainID)
+
+	addressLength, ok := args[1].(uint8)
+	if !ok || addressLength == 0 {
+		return nil, fmt.Errorf(exocmn.ErrContractInputParaOrType, 1, "uint8", args[1])
+	}
+	if addressLength < types.MinClientChainAddrLength {
+		return nil, xerrors.Errorf(exocmn.ErrInvalidAddrLength, addressLength, types.MinClientChainAddrLength)
+	}
+	clientChain.AddressLength = uint32(addressLength)
+
+	name, ok := args[2].(string)
+	if !ok {
+		return nil, xerrors.Errorf(exocmn.ErrContractInputParaOrType, 2, "string", args[2])
+	}
+	if name == "" || len(name) > types.MaxChainTokenNameLength {
+		return nil, xerrors.Errorf(exocmn.ErrInvalidNameLength, name, len(name), types.MaxChainTokenNameLength)
+	}
+	clientChain.Name = name
+
+	metaInfo, ok := args[3].(string)
+	if !ok {
+		return nil, xerrors.Errorf(exocmn.ErrContractInputParaOrType, 3, "string", args[2])
+	}
+	if metaInfo == "" || len(metaInfo) > types.MaxChainTokenMetaInfoLength {
+		return nil, xerrors.Errorf(exocmn.ErrInvalidMetaInfoLength, metaInfo, len(metaInfo), types.MaxChainTokenMetaInfoLength)
+	}
+	clientChain.MetaInfo = metaInfo
+
+	signatureType, ok := args[4].(string)
+	if !ok {
+		return nil, xerrors.Errorf(exocmn.ErrContractInputParaOrType, 4, "string", args[4])
+	}
+	clientChain.SignatureType = signatureType
+
+	return &clientChain, nil
+}
+
+func (p Precompile) TokenFromInputs(ctx sdk.Context, args []interface{}) (types.AssetInfo, error) {
+	if len(args) != 6 {
+		return types.AssetInfo{}, xerrors.Errorf(cmn.ErrInvalidNumberOfArgs, 6, len(args))
+	}
+	asset := types.AssetInfo{}
+	clientChainID, ok := args[0].(uint32)
+	if !ok {
+		return types.AssetInfo{}, fmt.Errorf(exocmn.ErrContractInputParaOrType, 0, "uint32", args[0])
+	}
+	asset.LayerZeroChainID = uint64(clientChainID)
+	info, err := p.assetsKeeper.GetClientChainInfoByIndex(ctx, asset.LayerZeroChainID)
+	if err != nil {
+		return types.AssetInfo{}, err
+	}
+	clientChainAddrLength := info.AddressLength
+
+	assetAddr, ok := args[1].([]byte)
```

Comment:
in`ExocoreGateway` I check that the asset address is non-zero, do you think it is needed here?

Reply:
The length of assetAddr will be checked to ensure it is not smaller than 20, so it is fine whether or not it is checked in the `ExocoreGateway`.

Reply:
this is different: even if the `assetAddr` is `address(0)`, the length of `assetAddr` would be 20

Reply:
I see your point. It doesn't seem necessary to check these special addresses here. If we are checking address(0), what about other special addresses like address(1) or address(0xffff)? It might be better to include these checks when registering the assets, even if we intend to address them. I'm not sure if checking these special addresses during DApp development is a default specification. If it is, adding related address checks during registration might be more standard.

Reply:
ok, we can skip this as not neccessary